GPflow/gplvm.py

```diff
@@ -0,0 +1,187 @@
+    to Q.
+    :param X: data array of size N (number of points) x D (dimensions)
+    :param Q: Number of latent dimensions, Q < D
+    :return: PCA projection array of size N x Q.
+    """
+    assert Q <= X.shape[1], 'Cannot have more latent dimensions than observed'
+    evecs, evals = np.linalg.eigh(np.cov(X.T))
+    i = np.argsort(evecs)[::-1]
+    W = evals[:, i]
+    W = W[:, :Q]
+    return (X - X.mean(0)).dot(W)
+
+
+class GPLVM(GPR):
+    """
+    Standard GPLVM where the likelihood can be optimised with respect to the latent X.
+    """
+
+    def __init__(self, Y, latent_dim, X_mean=None, kern=None, mean_function=Zero()):
+        """
+        Initialise GPLVM object. This method only works with a Gaussian likelihood.
+        :param Y: data matrix, size N (number of points) x D (dimensions)
+        :param Z: matrix of inducing points, size M (inducing points) x Q (latent dimensions)
+        :param X_mean: latent positions (N x Q), for the initialisation of the latent space.
+        :param kern: kernel specification, by default RBF
+        :param mean_function: mean function, by default None.
+        """
+        if kern is None:
+            kern = kernels.RBF(latent_dim, ARD=True)
+        if X_mean is None:
+            X_mean = PCA_reduce(Y, latent_dim)
+        assert X_mean.shape[1] == latent_dim, \
+            'Passed in number of latent ' + str(latent_dim) + ' does not match initial X ' + str(X_mean.shape[1])
+        self.num_latent = X_mean.shape[1]
+        assert Y.shape[1] >= self.num_latent, 'More latent dimensions than observed.'
+        GPR.__init__(self, X_mean, Y, kern, mean_function=mean_function)
+        del self.X  # in GPLVM this is a Param
+        self.X = Param(X_mean)
+
+
+class BayesianGPLVM(GPModel):
+    def __init__(self, X_mean, X_var, Y, kern, M, Z=None, X_prior_mean=None, X_prior_var=None):
+        """
+        Initialise Bayesian GPLVM object. This method only works with a Gaussian likelihood.
+        :param X_mean: initial latent positions, size N (number of points) x Q (latent dimensions).
+        :param X_var: variance of latent positions (N x Q), for the initialisation of the latent space.
+        :param Y: data matrix, size N (number of points) x D (dimensions)
+        :param kern: kernel specification, by default RBF
+        :param M: number of inducing points
+        :param Z: matrix of inducing points, size M (inducing points) x Q (latent dimensions). By default
+        random permutation of X_mean.
+        :param X_prior_mean: prior mean used in KL term of bound. By default 0. Same size as X_mean.
+        :param X_prior_var: pripor variance used in KL term of bound. By default 1.
+        """
+        GPModel.__init__(self, X_mean, Y, kern, likelihood=likelihoods.Gaussian(), mean_function=Zero())
+        del self.X  # in GPLVM this is a Param
+        self.X_mean = Param(X_mean)
+        # diag_transform = transforms.DiagMatrix(X_var.shape[1])
```

Comment:
Previously, the kernel expectations required rank 3 tensors, so I had to use a transform to constrain the NxD matrix of variances to a NxDxD tensor of diagonal matrices.

I've since reworked the kernel expectations to accept rank 2 tensors, and assume that this implies a diagonal q(X). This is a future-proof method for doing this, as we will want to use this representation to speed up the computation of the kernexps when q(X) is diagonal.